python/cudf/cudf/core/_base_index.py

```diff
@@ -608,19 +608,37 @@ def intersection(self, other, sort=False):
                     (1, 'Blue')],
                 )
         """
+        if not can_convert_to_column(other):
+            raise TypeError("Input must be Index or array-like")
+
         if not isinstance(other, BaseIndex):
-            other = cudf.Index(other, name=self.name)
+            other = cudf.Index(
+                other,
+                name=getattr(other, "name", self.name),
+            )
 
         if sort not in {None, False}:
             raise ValueError(
                 f"The 'sort' keyword only takes the values of "
                 f"None or False; {sort} was passed."
             )
 
-        if self.equals(other):
+        if not len(self) or self.equals(other):
+            dtypes = [self.dtype, other.dtype]
+            common_dtype = cudf.utils.dtypes.common_dtype_compatible(dtypes)
+
             if self.has_duplicates:
```

Comment:
Rather than duplicating this logic in the `if` statement body, can we write:
```python
result = self.unique if self.has_duplicates else self
return result._get_reconciled_name_object(other).astype(common_dtype)
```

Reply:
...or refactor this further to eliminate the same repetition in the `elif not len(other)` branch below.

Reply:
Refactored to remove code duplication.